File: volume1/chapter8/go_s3_example/s3_example.go
package main

import (
    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/s3"
    "github.com/aws/aws-sdk-go/service/s3/s3manager"    
    "fmt"
    "os"
    "path/filepath"
)

func main() {
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String("us-east-1")},
	)
	if (err != nil) {
		fmt.Println("Error", err)
		os.Exit(1)
	}

	// Create S3 service client
	svc := s3.New(sess)
	result, err := svc.ListBuckets(nil)
	bucket_name := "test1.replacewithyourdomain.com"
    	// range returns index, bucket pairs. We do not need the index
    	// so, we use the blank identifier _.
    	bucket_exists := false
	for _, b := range result.Buckets {
        bname := aws.StringValue(b.Name)
        if bname == bucket_name {
            bucket_exists = true
            break
        }
	}
    	bucket_exists = false
       if !bucket_exists {
           // Create bucket
           _, err = svc.CreateBucket(&s3.CreateBucketInput{
                    Bucket: aws.String(bucket_name),
                    })
           if err != nil {
               fmt.Printf("Unable to create bucket %s, %v", bucket_name, err)
           }        
       }
    
       // Upload file.
    svc_upload := s3manager.NewUploader(sess)    
    path_name := "path_to_your_file"
    file_name := filepath.Base(path_name)  
    fmt.Printf("file_name=%s", file_name)
	file, err := os.Open(path_name)
	if err != nil {
		fmt.Println("Failed to open file", path_name, err)
		os.Exit(1)
	}
	defer file.Close()  
    
	_, err_upload := svc_upload.Upload(&s3manager.UploadInput{
		Bucket: aws.String(bucket_name),
		Key:    aws.String(file_name),
		Body:   file,
	})
	if err_upload != nil {
		fmt.Println("error", err_upload)
		os.Exit(1)
	}    	
}

File: volume1/chapter8/nodejs_s3_example/s3_example.js
// Load the AWS SDK for Node.js
var AWS = require('aws-sdk');
// Set the region 
AWS.config.update({region: 'us-east-1'});

// Create S3 service object
s3 = new AWS.S3({apiVersion: '2006-03-01'});

// Call S3 to list the buckets
s3.listBuckets(function(err, data) {
  if (err) {
    console.log("Error", err);
  } else {
    var bucket_exists = false;
    for (var i = 0; i < data.Buckets.length; ++i) {
        var bucket = data.Buckets[i].Name;
        if (bucket.includes("test1.replacewithyourdomain.com")) {
           bucket_exists = true;
           break;
        }
    }
    if (!bucket_exists) {
        // Create bucket.
        var bucket_params = {
          Bucket : "test1.replacewithyourdomain.com",
          ACL : 'public-read'
        };

        // Call S3 to create the bucket
        s3.createBucket(bucket_params, function(err, data) {
          if (err) {
            console.log("Error", err);
          } else {
            console.log("Success", data.Location);
          }
        });
    }
    
    // Call S3 to upload file.
    //
    
    // Setup the file stream
    var file_system = require('fs');
    // Replace file with pathname of actual file.
    var file = "path_to_your_file";
    var file_stream = file_system.createReadStream(file);
    file_stream.on('error', function(err) {
      console.log('File Error', err);
    });    
    
    var path = require('path');    
    var key = path.basename(file);
    var body = file_stream;    
    var upload_params = {
        Bucket: "test1.replacewithyourdomain.com", 
        Key: key, 
        Body: body
    };

    // Call S3.upload() to upload file.
    s3.upload (upload_params, function (err, data) {
      if (err) {
        console.log("Error", err);
      } if (data) {
        console.log("Upload Success", data.Location);
      }
    });    
 
  }
});

